File: src/Gameplay/House/House.js
"use strict";
/// <reference path='../../../types-gt-mp/Definitions/index.d.ts' />
var house_text_mode = 0;
var remove_selected_furniture = false;
var house_purchase_menu = null;
var house_menu = null;
var house_safe_menu = null;
var house_guns_main_menu = null;
var house_guns_locker_menu = null;
var house_furnitures_main_menu = null;
var house_furnitures_buy_menu = null;
var house_furnitures_menu = null;
var house_menus = [];
function h_numberWithCommas(x) {
    return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}
function hideMenus() {
    for (var i = 0; i < house_menus.length; i++)
        house_menus[i].Visible = false;
}
API.onServerEventTrigger.connect(function (event_name, args) {
    switch (event_name) {
        case "ShowHouseText":
            house_text_mode = args[0];
            if (house_text_mode == 0)
                hideMenus();
            break;
        case "House_PurchaseMenu":
            var data = JSON.parse(args[0]);
            if (house_purchase_menu == null) {
                house_purchase_menu = API.createMenu("Casa", "~g~Esta casa está a venda!", 0, 0, 6);
                house_menus.push(house_purchase_menu);
                house_purchase_menu.OnItemSelect.connect(function (menu, item, index) {
                    if (index == 2)
                        API.triggerServerEvent("HousePurchase");
                });
            }
            house_purchase_menu.Clear();
            var temp_item = API.createMenuItem("Interior", "Interior da casa.");
            temp_item.SetRightLabel(data.Interior);
            house_purchase_menu.AddItem(temp_item);
            temp_item = API.createMenuItem("Preço", "Preço da casa.");
            temp_item.SetRightLabel("$" + h_numberWithCommas(data.Price));
            house_purchase_menu.AddItem(temp_item);
            temp_item = API.createColoredItem("Comprar", "Comprar a casa.", "#4caf50", "#ffffff");
            house_purchase_menu.AddItem(temp_item);
            house_purchase_menu.Visible = true;
            break;
        case "HouseMenu":
            var data = JSON.parse(args[0]);
            if (house_menu == null) {
                house_menu = API.createMenu("Casa", "~b~Controles da casa", 0, 0, 6);
                house_menus.push(house_menu);
                house_menu.OnCheckboxChange.connect(function (menu, item, checked) {
                    API.triggerServerEvent("HouseSetLock", checked);
                });
                house_menu.OnItemSelect.connect(function (menu, item, index) {
                    switch (index) {
                        case 0:
                            var name = API.getUserInput("", 31);
                            if (name.length > 0)
                                API.triggerServerEvent("HouseSetName", name);
                            break;
                        case 5:
                            API.triggerServerEvent("HouseSell");
                            break;
                    }
                });
            }
            if (house_safe_menu == null) {
                house_safe_menu = API.createMenu("Casa", "~b~Cofre", 0, 0, 6);
                house_menus.push(house_safe_menu);
                house_safe_menu.ParentMenu = house_menu;
                var temp_safe_item = API.createMenuItem("Depositar", "Coloca dinheiro na casa.");
                house_safe_menu.AddItem(temp_safe_item);
                temp_safe_item = API.createMenuItem("Sacar", "Retira dinheiro da casa.");
                house_safe_menu.AddItem(temp_safe_item);
                house_safe_menu.OnItemSelect.connect(function (menu, item, index) {
                    var amount = API.getUserInput("", 9);
                    if (parseInt(amount) > 1)
                        API.triggerServerEvent("HouseSafe", index, amount);
                });
            }
            if (house_guns_main_menu == null) {
                house_guns_main_menu = API.createMenu("Casa", "~b~Armário de armas", 0, 0, 6);
                house_menus.push(house_guns_main_menu);
                house_guns_locker_menu = API.createMenu("Casa", "~b~Armário de armas (Armazenado)", 0, 0, 6);
                house_menus.push(house_guns_locker_menu);
                house_guns_main_menu.ParentMenu = house_menu;
                house_guns_locker_menu.ParentMenu = house_guns_main_menu;
                var temp_guns_item = API.createMenuItem("Colocar arma", "Coloca sua arma atual no armário.");
                house_guns_main_menu.AddItem(temp_guns_item);
                temp_guns_item = API.createMenuItem("Retirar arma", "Retira uma arma do armário.");
                house_guns_main_menu.AddItem(temp_guns_item);
                house_guns_main_menu.BindMenuToItem(house_guns_locker_menu, temp_guns_item);
                house_guns_main_menu.OnItemSelect.connect(function (menu, item, index) {
                    if (index == 0)
                        API.triggerServerEvent("HousePutGun");
                });
                house_guns_locker_menu.OnItemSelect.connect(function (menu, item, index) {
                    API.triggerServerEvent("HouseTakeGun", index);
                });
            }
            if (house_furnitures_main_menu == null) {
                house_furnitures_main_menu = API.createMenu("Casa", "~b~Móveis", 0, 0, 6);
                house_menus.push(house_furnitures_main_menu);
                house_furnitures_buy_menu = API.createMenu("Casa", "~b~Catálogo de móveis", 0, 0, 6);
                house_menus.push(house_furnitures_buy_menu);
                house_furnitures_menu = API.createMenu("Casa", "~b~Móveis", 0, 0, 6);
                house_menus.push(house_furnitures_menu);
                house_furnitures_main_menu.ParentMenu = house_menu;
                house_furnitures_buy_menu.ParentMenu = house_furnitures_main_menu;
                house_furnitures_menu.ParentMenu = house_furnitures_main_menu;
                var temp_furnitures_item = API.createMenuItem("Comprar móveis", "Compra móveis.");
                house_furnitures_main_menu.AddItem(temp_furnitures_item);
                house_furnitures_main_menu.BindMenuToItem(house_furnitures_buy_menu, temp_furnitures_item);
                temp_furnitures_item = API.createMenuItem("Editar móvel", "Seleciona um móvel para editar.");
                house_furnitures_main_menu.AddItem(temp_furnitures_item);
                house_furnitures_main_menu.BindMenuToItem(house_furnitures_menu, temp_furnitures_item);
                temp_furnitures_item = API.createMenuItem("Vender móvel", "Seleciona um móvel para vender.");
                house_furnitures_main_menu.AddItem(temp_furnitures_item);
                house_furnitures_main_menu.BindMenuToItem(house_furnitures_menu, temp_furnitures_item);
                house_furnitures_main_menu.OnItemSelect.connect(function (menu, item, index) {
                    switch (index) {
                        case 0:
                            API.triggerServerEvent("HouseFurnitureCatalogue");
                            break;
                        case 1:
                            remove_selected_furniture = false;
                            break;
                        case 2:
                            remove_selected_furniture = true;
                            break;
                    }
                });
                house_furnitures_buy_menu.OnItemSelect.connect(function (menu, item, index) {
                    API.triggerServerEvent("HouseBuyFurniture", index);
                });
                house_furnitures_menu.OnItemSelect.connect(function (menu, item, index) {
                    API.triggerServerEvent(((remove_selected_furniture) ? "HouseSellFurniture" : "HouseEditFurniture"), index);
                });
            }
            // house main menu
            house_menu.Clear();
            var temp_item = API.createMenuItem("Nome da casa", "Altera o nome de sua casa.");
            house_menu.AddItem(temp_item);
            temp_item = API.createCheckboxItem("Trancada", "Tranca/destranca a sua casa.", data.Locked);
            house_menu.AddItem(temp_item);
            temp_item = API.createMenuItem("Cofre", "Opções do cofre.");
            house_menu.AddItem(temp_item);
            house_menu.BindMenuToItem(house_safe_menu, temp_item);
            temp_item = API.createMenuItem("Armário de armas", "Opções do armário de armas.");
            house_menu.AddItem(temp_item);
            house_menu.BindMenuToItem(house_guns_main_menu, temp_item);
            temp_item = API.createMenuItem("Móveis", "Opções dos móveis.");
            house_menu.AddItem(temp_item);
            house_menu.BindMenuToItem(house_furnitures_main_menu, temp_item);
            temp_item = API.createColoredItem("Vender casa", "Vende a casa.", "#4caf50", "#ffffff");
            house_menu.AddItem(temp_item);
            // house safe menu
            API.setMenuSubtitle(house_safe_menu, "~b~Cofre ~g~($" + h_numberWithCommas(data.Money) + ")");
            // house guns menu
            house_guns_locker_menu.Clear();
            for (var i = 0; i < data.Weapons.length; i++) {
                var temp_locker_item = API.createMenuItem(API.getWeaponName(data.Weapons[i].Hash), "");
                temp_locker_item.SetRightLabel("Munição: " + h_numberWithCommas(data.Weapons[i].Ammo));
                temp_locker_item.Description = ((data.Weapons[i].Components.length > 0) ? "Tem componentes." : "Não tem componentes.");
                if (data.Weapons[i].Tint > 0)
                    temp_locker_item.Description += " (Tingida)";
                house_guns_locker_menu.AddItem(temp_locker_item);
            }
            // house furnitures menu
            house_furnitures_menu.Clear();
            for (var i = 0; i < data.Furnitures.length; i++) {
                var temp_furniture_item = API.createMenuItem(data.Furnitures[i].Name, "");
                temp_furniture_item.SetRightLabel("$" + h_numberWithCommas(Math.round(data.Furnitures[i].Price * 0.8)));
                house_furnitures_menu.AddItem(temp_furniture_item);
            }
            house_menu.Visible = true;
            break;
        case "HouseUpdateSafe":
            if (house_safe_menu != null) {
                var data = JSON.parse(args[0]);
                API.setMenuSubtitle(house_safe_menu, "~b~Cofre ~g~($" + h_numberWithCommas(data.Money) + ")");
            }
            break;
        case "HouseUpdateWeapons":
            if (house_guns_locker_menu != null) {
                var data = JSON.parse(args[0]);
                house_guns_locker_menu.Clear();
                for (var i = 0; i < data.length; i++) {
                    var temp_locker_item = API.createMenuItem(API.getWeaponName(data[i].Hash), "");
                    temp_locker_item.SetRightLabel("Munição: " + h_numberWithCommas(data[i].Ammo));
                    temp_locker_item.Description = ((data[i].Components.length > 0) ? "Tem componentes." : "Não tem componentes.");
                    if (data[i].Tint > 0)
                        temp_locker_item.Description += " (Tingida)";
                    house_guns_locker_menu.AddItem(temp_locker_item);
                }
            }
            break;
        case "HouseFurnitureCatalogue":
            if (house_furnitures_buy_menu != null) {
                var data = JSON.parse(args[0]);
                house_furnitures_buy_menu.Clear();
                for (var i = 0; i < data.length; i++) {
                    var temp_furniture_item = API.createMenuItem(data[i].Name, "");
                    temp_furniture_item.SetRightLabel("$" + h_numberWithCommas(data[i].Price));
                    house_furnitures_buy_menu.AddItem(temp_furniture_item);
                }
            }
            break;
        case "HouseUpdateFurnitures":
            if (house_furnitures_menu != null) {
                var data = JSON.parse(args[0]);
                house_furnitures_menu.Clear();
                for (var i = 0; i < data.length; i++) {
                    var temp_furniture_item = API.createMenuItem(data[i].Name, "");
                    temp_furniture_item.SetRightLabel("$" + h_numberWithCommas(Math.round(data[i].Price * 0.8)));
                    house_furnitures_menu.AddItem(temp_furniture_item);
                }
            }
            break;
        case "UpdateHouseBlip":
            API.setBlipColor(args[0], 69);
            API.setBlipShortRange(args[0], false);
            break;
        case "ResetHouseBlip":
            API.setBlipColor(args[0], 0);
            API.setBlipShortRange(args[0], true);
            break;
    }
});
API.onKeyDown.connect(function (e, key) {
    if (API.isChatOpen())
        return;
    switch (key.KeyCode) {
        case Keys.E:
            if (resource.FurnitureEditor.editing_handle != null)
                return;
            if (house_text_mode == 1) {
                API.triggerServerEvent("HouseInteract");
            }
            else if (house_text_mode == 2) {
                API.triggerServerEvent("HouseLeave");
            }
            break;
        case Keys.M:
            if (resource.FurnitureEditor.editing_handle != null)
                return;
            hideMenus();
            API.triggerServerEvent("HouseMenu");
            break;
    }
});
API.onEntityStreamIn.connect(function (ent, entType) {
    if (entType === 3) {
        var count = API.getEntitySyncedData(ent, "PlayersInside");
        if (count < 1) {
            API.callNative("HIDE_NUMBER_ON_BLIP", ent);
        }
        else {
            API.callNative("SHOW_NUMBER_ON_BLIP", ent, count);
        }
    }
});
API.onUpdate.connect(function () {
    if (house_text_mode > 0)
        API.displaySubtitle(((house_text_mode == 1) ? "Pressione ~y~E ~w~para interagir." : "Pressione ~y~E ~w~para sair."), 100);
});

File: src/Gameplay/House/House.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using GrandTheftMultiplayer.Server.API;
using GrandTheftMultiplayer.Server.Constant;
using GrandTheftMultiplayer.Server.Elements;
using GrandTheftMultiplayer.Server.Managers;
using GrandTheftMultiplayer.Shared;
using GrandTheftMultiplayer.Shared.Math;
using pcrpg.src.Database.Models;
using Newtonsoft.Json;
using pcrpg.src.Player.Utils;

namespace pcrpg.src.Gameplay.House
{
    #region House Class
    public class House
    {
        public Guid ID { get; }
        public string Owner { get; private set; }
        public int Type { get; private set; }
        public Vector3 Position { get; }
        public int Price { get; private set; }
        public bool Locked { get; private set; }
        [JsonIgnore] public int Dimension { get; set; }

        // customization
        public string Name { get; private set; }
        public List<HouseFurniture> Furnitures { get; }

        // storage
        public long Money { get; private set; }
        public List<HouseWeapon> Weapons { get; }

        // entities
        [JsonIgnore]
        private Blip Blip;

        [JsonIgnore]
        private Marker Marker;

        [JsonIgnore]
        private ColShape ColShape;

        [JsonIgnore]
        private TextLabel Label;

        // misc
        [JsonIgnore]
        private List<NetHandle> PlayersInside = new List<NetHandle>();

        [JsonIgnore]
        private Client OwnerHandle;

        [JsonIgnore]
        private DateTime LastSave;

        public House(Guid id, string owner, int type, Vector3 position, int price, bool locked, string name = "", List<HouseFurniture> furnitures = null, long money = 0, List<HouseWeapon> weapons = null)
        {
            ID = id;
            Owner = owner;
            Type = type;
            Position = position;
            Price = price;
            Locked = locked;

            Name = name;
            Furnitures = (furnitures == null) ? new List<HouseFurniture>() : furnitures;

            Money = money;
            Weapons = (weapons == null) ? new List<HouseWeapon>() : weapons;

            // create blip
            Blip = API.shared.createBlip(position);

            if (string.IsNullOrEmpty(owner))
            {
                Blip.sprite = 350;
                Blip.name = "Casa a venda";
            }
            else
            {
                Blip.sprite = 40;
                Blip.name = "Casa de jogador";
            }

            Blip.color = 0;
            Blip.scale = 1f;
            Blip.shortRange = true;
            Blip.setSyncedData("PlayersInside", 0);

            // create marker
            Marker = API.shared.createMarker(1, position - new Vector3(0.0, 0.0, 1.0), new Vector3(), new Vector3(), new Vector3(1.0, 1.0, 0.5), 150, 64, 196, 255);

            // create colshape
            ColShape = API.shared.createCylinderColShape(position, 0.85f, 0.85f);
            ColShape.onEntityEnterColShape += (s, ent) =>
            {
                Client player;

                if ((player = API.shared.getPlayerFromHandle(ent)) != null)
                {
                    player.setData("HouseMarker_ID", ID);
                    player.triggerEvent("ShowHouseText", 1);
                }
            };

            ColShape.onEntityExitColShape += (s, ent) =>
            {
                Client player;

                if ((player = API.shared.getPlayerFromHandle(ent)) != null)
                {
                    player.resetData("HouseMarker_ID");
                    player.triggerEvent("ShowHouseText", 0);
                }
            };

            // create text label
            Label = API.shared.createTextLabel("House", position, 15f, 0.65f);
            UpdateLabel();

            OwnerHandle = null;
        }

        private void UpdateLabel()
        {
            if (string.IsNullOrEmpty(Owner))
            {
                Label.text = string.Format("~g~Casa a venda~n~~w~Interior: ~g~{0}~n~~w~Preço: ~g~${1:n0}", HouseTypes.HouseTypeList[Type].Name, Price);
            }
            else
            {
                Label.text = string.Format("~g~Casa~n~~w~{0}~n~~w~Proprietário: ~g~{1}~n~{2}", Name, Owner, ((Locked) ? "~r~Trancada" : "~g~Destrancada"));
            }
        }

        private void UpdateBlip()
        {
            int count = PlayersInside.Count;
            Blip.setSyncedData("PlayersInside", count);

            if (count < 1)
            {
                if (Blip.sprite != 40)
                {
                    Blip.sprite = 40;

                    if (OwnerHandle != null) UpdateBlipForOwner();
                }

                API.shared.sendNativeToAllPlayers(Hash.HIDE_NUMBER_ON_BLIP, Blip.handle);
            }
            else
            {
                if (Blip.sprite != 417)
                {
                    Blip.sprite = 417;

                    if (OwnerHandle != null) UpdateBlipForOwner();
                }

                API.shared.sendNativeToAllPlayers(Hash.SHOW_NUMBER_ON_BLIP, Blip.handle, count);
            }
        }

        public void SetOwner(Client player)
        {
            if (OwnerHandle != null) OwnerHandle.triggerEvent("ResetHouseBlip", Blip.handle);
            Owner = (player == null) ? string.Empty : player.name;

            Blip.sprite = (player == null) ? 350 : 40;
            Blip.color = 0;
            Blip.shortRange = true;

            SetOwnerHandle(player);
            UpdateLabel();
            Save();
        }

        public void SetOwnerHandle(Client player)
        {
            OwnerHandle = player;
            if (player != null) player.triggerEvent("UpdateHouseBlip", Blip.handle);
        }

        public void UpdateBlipForOwner()
        {
            if (OwnerHandle == null) return;
            OwnerHandle.triggerEvent("UpdateHouseBlip", Blip.handle);
        }

        public void SetName(string new_name)
        {
            Name = new_name;

            UpdateLabel();
            Save();
        }

        public void SetLock(bool locked)
        {
            Locked = locked;

            UpdateLabel();
            Save();
        }

        public void SetType(int new_type)
        {
            Type = new_type;

            UpdateLabel();
            Save();
        }

        public void SetPrice(int new_price)
        {
            Price = new_price;

            UpdateLabel();
            Save();
        }

        public void ChangeMoney(int amount)
        {
            Money += amount;

            Save();
        }

        public void SendPlayer(Client player)
        {
            player.position = HouseTypes.HouseTypeList[Type].Position;
            player.dimension = Dimension;
            player.setData("InsideHouse_ID", ID);

            if (!PlayersInside.Contains(player.handle)) PlayersInside.Add(player.handle);
            UpdateBlip();
        }

        public void RemovePlayer(Client player, bool set_pos = true)
        {
            if (set_pos)
            {
                player.position = Position;
                player.dimension = 0;
            }

            player.resetData("InsideHouse_ID");

            if (PlayersInside.Contains(player.handle)) PlayersInside.Remove(player.handle);
            UpdateBlip();
        }

        public void RemoveAllPlayers(bool exit = false)
        {
            for (int i = PlayersInside.Count - 1; i >= 0; i--)
            {
                Client player = API.shared.getEntityFromHandle<Client>(PlayersInside[i]);

                if (player != null)
                {
                    player.position = Position;
                    player.dimension = 0;

                    player.resetData("InsideHouse_ID");
                }

                PlayersInside.RemoveAt(i);
            }

            if (!exit) UpdateBlip();
        }

        public void Save(bool force = false)
        {
            if (!force && DateTime.Now.Subtract(LastSave).TotalSeconds < Main.SAVE_INTERVAL) return;

            File.WriteAllText(Main.HOUSE_SAVE_DIR + Path.DirectorySeparatorChar + ID + ".json", JsonConvert.SerializeObject(this, Formatting.Indented));
            LastSave = DateTime.Now;
        }

        public void Destroy(bool exit = false)
        {
            foreach (HouseFurniture furniture in Furnitures) furniture.Destroy();
            RemoveAllPlayers(exit);

            Blip.delete();
            Marker.delete();
            API.shared.deleteColShape(ColShape);
            Label.delete();
        }
    }
    #endregion

    public class Main : Script
    {
        // settings
        public static string HOUSE_SAVE_DIR = "data/House";
        public static int PLAYER_HOUSE_LIMIT = 3;
        public static int HOUSE_MONEY_LIMIT = 5000000;
        public static int HOUSE_WEAPON_LIMIT = 10;
        public static int HOUSE_FURNITURE_LIMIT = 25;
        public static bool RESET_DIMENSION_ON_DEATH = true;
        public static int SAVE_INTERVAL = 120;

        public static List<House> Houses = new List<House>();
        public static int DimensionID = 1;

        public Main()
        {
            API.onResourceStart += House_Init;

            API.onPlayerFinishedDownload += House_PlayerJoin;
            API.onClientEventTrigger += House_ClientEvent;
            API.onPlayerDeath += House_PlayerDeath;
            API.onPlayerDisconnected += House_PlayerLeave;

            API.onResourceStop += House_Exit;
        }

        #region Methods
        public static Guid GetGuid()
        {
            Guid new_guid;

            do
            {
                new_guid = Guid.NewGuid();
            } while (Houses.Count(h => h.ID == new_guid) > 0);

            return new_guid;
        }

        public void RemovePlayerFromHouseList(Client player)
        {
            if (player.hasData("InsideHouse_ID"))
            {
                House house = Houses.FirstOrDefault(h => h.ID == player.getData("InsideHouse_ID"));
                if (house == null) return;

                house.RemovePlayer(player, false);
            }
        }
        #endregion

        #region Events
        public void House_Init()
        {
            DimensionID = 1;

            // load settings
            if (API.hasSetting("houseDirName")) HOUSE_SAVE_DIR = API.getSetting<string>("houseDirName");

            HOUSE_SAVE_DIR = API.getResourceFolder() + Path.DirectorySeparatorChar + HOUSE_SAVE_DIR;
            if (!Directory.Exists(HOUSE_SAVE_DIR)) Directory.CreateDirectory(HOUSE_SAVE_DIR);

            if (API.hasSetting("playerHouseLimit")) PLAYER_HOUSE_LIMIT = API.getSetting<int>("playerHouseLimit");
            if (API.hasSetting("houseMoneyLimit")) HOUSE_MONEY_LIMIT = API.getSetting<int>("houseMoneyLimit");
            if (API.hasSetting("houseWeaponLimit")) HOUSE_WEAPON_LIMIT = API.getSetting<int>("houseWeaponLimit");
            if (API.hasSetting("houseFurnitureLimit")) HOUSE_FURNITURE_LIMIT = API.getSetting<int>("houseFurnitureLimit");
            if (API.hasSetting("resetDimensionOnDeath")) RESET_DIMENSION_ON_DEATH = API.getSetting<bool>("resetDimensionOnDeath");
            if (API.hasSetting("saveInterval")) SAVE_INTERVAL = API.getSetting<int>("saveInterval");

            API.consoleOutput("-> Player House Limit: {0}", ((PLAYER_HOUSE_LIMIT == 0) ? "Disabled" : PLAYER_HOUSE_LIMIT.ToString()));
            API.consoleOutput("-> House Safe Limit: ${0:n0}", HOUSE_MONEY_LIMIT);
            API.consoleOutput("-> House Weapon Limit: {0}", ((HOUSE_WEAPON_LIMIT == 0) ? "Disabled" : HOUSE_WEAPON_LIMIT.ToString()));
            API.consoleOutput("-> House Furniture Limit: {0}", ((HOUSE_FURNITURE_LIMIT == 0) ? "Disabled" : HOUSE_FURNITURE_LIMIT.ToString()));
            API.consoleOutput("-> Dimension Reset On Death: {0}", ((RESET_DIMENSION_ON_DEATH) ? "Enabled" : "Disabled"));

            // load houses
            foreach (string file in Directory.EnumerateFiles(HOUSE_SAVE_DIR, "*.json"))
            {
                House house = JsonConvert.DeserializeObject<House>(File.ReadAllText(file));
                house.Dimension = DimensionID;
                foreach (HouseFurniture furniture in house.Furnitures) furniture.Create(DimensionID);

                Houses.Add(house);
                DimensionID++;
            }

            API.consoleOutput("Loaded {0} houses.", Houses.Count);
        }

        public void House_PlayerJoin(Client player)
        {
            API.delay(2000, true, () => { foreach (House house in Houses.Where(h => h.Owner == player.name)) house.SetOwnerHandle(player); });
        }

        public void House_ClientEvent(Client player, string event_name, params object[] args)
        {
            switch (event_name)
            {
                case "HouseInteract":
                    {
                        if (!player.hasData("HouseMarker_ID")) return;

                        House house = Houses.FirstOrDefault(h => h.ID == player.getData("HouseMarker_ID"));
                        if (house == null) return;

                        if (string.IsNullOrEmpty(house.Owner))
                        {
                            // not owned house
                            player.triggerEvent("House_PurchaseMenu", API.toJson(new { Interior = HouseTypes.HouseTypeList[house.Type].Name, Price = house.Price }));
                        }
                        else
                        {
                            // owned house
                            if (house.Locked)
                            {
                                if (house.Owner == player.name)
                                {
                                    house.SendPlayer(player);
                                }
                                else
                                {
                                    player.sendNotification("Error", "~r~Apenas o proprietário pode acessar esta casa.");
                                }
                            }
                            else
                            {
                                house.SendPlayer(player);
                            }
                        }

                        break;
                    }

                case "HousePurchase":
                    {
                        if (!player.hasData("HouseMarker_ID")) return;

                        House house = Houses.FirstOrDefault(h => h.ID == player.getData("HouseMarker_ID"));
                        if (house == null) return;

                        if (!string.IsNullOrEmpty(house.Owner))
                        {
                            player.sendNotification("Error", "~r~Esta casa já tem dono.");
                            return;
                        }

                        if (house.Price > player.getMoney())
                        {
                            player.sendNotification("Error", "~r~Você não tem dinheiro suficiente.");
                            return;
                        }

                        if (PLAYER_HOUSE_LIMIT > 0 && Houses.Count(h => h.Owner == player.name) >= PLAYER_HOUSE_LIMIT)
                        {
                            player.sendNotification("Error", "~r~Você não pode ter mais casas.");
                            return;
                        }

                        player.sendNotification("House Purchased", "~g~Parabéns, você comprou a casa!");

                        house.SetLock(true);
                        house.SetOwner(player);
                        house.SendPlayer(player);

                        player.giveMoney(-house.Price);
                        break;
                    }

                case "HouseMenu":
                    {
                        if (!player.hasData("InsideHouse_ID")) return;

                        House house = Houses.FirstOrDefault(h => h.ID == player.getData("InsideHouse_ID"));
                        if (house == null) return;

                        if (house.Owner != player.name)
                        {
                            player.sendNotification("Error", "~r~Apenas o proprietário pode acessar o menu da casa.");
                            return;
                        }

                        player.triggerEvent("HouseMenu", API.toJson(house));
                        break;
                    }

                case "HouseSetName":
                    {
                        if (!player.hasData("InsideHouse_ID") || args.Length < 1) return;

                        House house = Houses.FirstOrDefault(h => h.ID == player.getData("InsideHouse_ID"));
                        if (house == null) return;

                        if (house.Owner != player.name)
                        {
                            player.sendNotification("Error", "~r~Apenas o dono pode fazer isto.");
                            return;
                        }

                        string new_name = args[0].ToString();
                        if (new_name.Length > 32)
                        {
                            player.sendNotification("Error", "~r~O nome da casa não pode ter mais que 32 caracteres.");
                            return;
                        }

                        house.SetName(new_name);
                        player.sendNotification("Success", string.Format("~g~Nome da casa alterado para: ~w~\"{0}\"", new_name));
                        break;
                    }

                case "HouseSetLock":
                    {
                        if (!player.hasData("InsideHouse_ID") || args.Length < 1) return;

                        House house = Houses.FirstOrDefault(h => h.ID == player.getData("InsideHouse_ID"));
                        if (house == null) return;

                        if (house.Owner != player.name)
                        {
                            player.sendNotification("Error", "~r~Apenas o dono pode fazer isto.");
                            return;
                        }

                        bool new_state = Convert.ToBoolean(args[0]);
                        house.SetLock(new_state);

                        player.sendNotification("Success", ((new_state) ? "~g~A casa agora está trancada." : "~g~A casa agora está destrancada."));
                        break;
                    }

                case "HouseSafe":
                    {
                        if (!player.hasData("InsideHouse_ID") || args.Length < 2) return;

                        House house = Houses.FirstOrDefault(h => h.ID == player.getData("InsideHouse_ID"));
                        if (house == null) return;

                        if (house.Owner != player.name)
                        {
                            player.sendNotification("Error", "~r~Apenas o dono pode fazer isto.");
                            return;
                        }

                        int type = Convert.ToInt32(args[0]);
                        int amount = 0;

                        if (!int.TryParse(args[1].ToString(), out amount))
                        {
                            player.sendNotification("Error", "~r~Quantidade inválida.");
                            return;
                        }

                        if (amount < 1) return;
                        if (type == 0)
                        {
                            if (player.getMoney() < amount)
                            {
                                player.sendNotification("Error", "~r~Você não tem dinheiro suficiente.");
                                return;
                            }

                            if (house.Money + amount > HOUSE_MONEY_LIMIT)
                            {
                                player.sendNotification("Error", "~r~O cofre está cheio.");
                                return;
                            }

                            player.giveMoney(-amount);

                            house.ChangeMoney(amount);
                            player.sendNotification("Success", string.Format("~g~Você colocou ${0:n0} no cofre.", amount));
                            player.triggerEvent("HouseUpdateSafe", API.toJson(new { Money = house.Money }));
                        }
                        else
                        {
                            if (house.Money < amount)
                            {
                                player.sendNotification("Error", "~r~O cofre não possui esta quantia.");
                                return;
                            }

                            player.giveMoney(amount);

                            house.ChangeMoney(-amount);
                            player.sendNotification("Success", string.Format("~g~Você tirou ${0:n0} do cofre.", amount));
                            player.triggerEvent("HouseUpdateSafe", API.toJson(new { Money = house.Money }));
                        }

                        break;
                    }

                case "HouseSell":
                    {
                        if (!player.hasData("InsideHouse_ID")) return;

                        House house = Houses.FirstOrDefault(h => h.ID == player.getData("InsideHouse_ID"));
                        if (house == null) return;

                        if (house.Owner != player.name)
                        {
                            player.sendNotification("Error", "~r~Apenas o dono pode fazer isto.");
                            return;
                        }

                        if (house.Money > 0)
                        {
                            player.sendNotification("Error", "~r~Esvazie o cofre antes de vender.");
                            return;
                        }

                        if (house.Weapons.Count > 0)
                        {
                            player.sendNotification("Error", "~r~Esvazie o armário de armas antes de vender.");
                            return;
                        }

                        if (house.Furnitures.Count > 0)
                        {
                            player.sendNotification("Error", "~r~Venda os móveis antes de vender.");
                            return;
                        }

                        int price = (int)Math.Round(house.Price * 0.8);
                        player.giveMoney(price);

                        house.RemoveAllPlayers();
                        house.SetOwner(null);

                        player.sendNotification("Success", string.Format("~g~Casa vendida por ${0:n0}.", price));
                        break;
                    }

                case "HouseLeave":
                    {
                        if (!player.hasData("InsideHouse_ID")) return;

                        House house = Houses.FirstOrDefault(h => h.ID == player.getData("InsideHouse_ID"));
                        if (house == null) return;

                        house.RemovePlayer(player);
                        break;
                    }
            }
        }

        public void House_PlayerDeath(Client player, NetHandle killer, int weapon)
        {
            if (RESET_DIMENSION_ON_DEATH) player.dimension = 0;
            RemovePlayerFromHouseList(player);
        }

        public void House_PlayerLeave(Client player, string reason)
        {
            RemovePlayerFromHouseList(player);
            foreach (House house in Houses.Where(h => h.Owner == player.name)) house.SetOwnerHandle(null);
        }

        public void House_Exit()
        {
            foreach (House house in Houses)
            {
                house.Save(true);
                house.Destroy(true);
            }

            Houses.Clear();
        }
        #endregion
    }
}
